### PR TITLE
First draft of Monitoring plugin feature.

### PR DESCRIPTION
**Pull request by idanl21**

### CODE REVIEW

File: opensearch-operator/pkg/builders/cluster.go [hunk 1]
```diff
@@ -852,3 +859,56 @@ func DataNodesCount(ctx context.Context, k8sClient client.Client, cr *opsterv1.O
...
+				{Port: "9200",
+					TargetPort:      nil,
+					Path:            "/_prometheus/metrics",
+					Interval:        "10s",
+					TLSConfig:       nil,
+					BearerTokenFile: "",
+					HonorLabels:     false,
+					BasicAuth: &prometheus.BasicAuth{
+						Username: corev1.SecretKeySelector{
+							LocalObjectReference: corev1.LocalObjectReference{Name: cr.ClusterName + "-admin-password "},
```
Comment (swoehrl-mw) [inline opensearch-operator/pkg/builders/cluster.go:900]: Not sure how I feel about using the admin password for this, I would prefer a separate user with limited permissions. But that would be hard to implement universally.
Suggestion: Let the user optionally provide credentials for monitoring. If they are not provided, fall back to admin creds as you implemented it now.

File: opensearch-operator/pkg/builders/cluster.go [hunk 2]
```diff
@@ -852,3 +859,56 @@ func DataNodesCount(ctx context.Context, k8sClient client.Client, cr *opsterv1.O
...
+				cr.ClusterName,
+			},
+			PodTargetLabels: []string{
+				cr.ClusterName,
+			},
+			Endpoints: []prometheus.Endpoint{
+				{Port: "9200",
+					TargetPort:      nil,
+					Path:            "/_prometheus/metrics",
+					Interval:        "10s",
```
Comment (swoehrl-mw) [inline opensearch-operator/pkg/builders/cluster.go:894]: I think `10s` is too often. From my perspective `30s` is a good default. Maybe make this configurable for the user?

File: opensearch-operator/api/v1/opensearch_types.go [hunk 3]
```diff
@@ -47,6 +47,7 @@ type GeneralConfig struct {
 	PluginsList    []string `json:"pluginsList,omitempty"`
 	// Additional volumes to mount to all pods in the cluster
 	AdditionalVolumes []AdditionalVolume `json:"additionalVolumes,omitempty"`
+	Monitoring        bool               `json:"monitoring,omitempty"`
```
Comment (swoehrl-mw) [inline opensearch-operator/api/v1/opensearch_types.go:50]: I would prefer for this to be an object where one field is a boolean `enabled` field. Gives us more flexibility to make the monitoring configurable by the user.

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]
Hi @idanl21 . I left you some comments.
In addition
* I'm missing a unittest for this feature
* Documentation in the userguide is missing
* The new field is not documented in `docs/designs/crd.md`
* The CRD yaml (both in `config/crd/bases` and in the helm chart) is not updated

File: opensearch-operator/pkg/builders/cluster.go [hunk 4]
```diff
@@ -193,6 +194,12 @@ func NewSTSForNodePool(
 
 	var mainCommand []string
 	com := "./bin/opensearch-plugin install --batch"
+	if cr.Spec.General.Monitoring {
+		clusterVersion := cr.Spec.General.Version
+		url := "https://github.com/aiven/prometheus-exporter-plugin-for-opensearch/releases/download/" + clusterVersion + ".0/prometheus-exporter" + clusterVersion + ".0.zip"
```
Comment (prudhvigodithi) [inline opensearch-operator/pkg/builders/cluster.go:199]: Hey @idanl21 will this work for operator offline setup?, Like when operator installed in a restrictive environment with limited internet URL access, can it then pull this plugin?

File: opensearch-operator/pkg/builders/cluster.go (hunk 1 above)
Comment (idanl21) [reply opensearch-operator/pkg/builders/cluster.go:900]: Good idea, i will implement it.
the user should be already created? or we will create it as part of the flow?

File: opensearch-operator/pkg/builders/cluster.go (hunk 2 above)
Comment (idanl21) [reply opensearch-operator/pkg/builders/cluster.go:894]: Will make it configurable

File: opensearch-operator/api/v1/opensearch_types.go (hunk 3 above)
Comment (idanl21) [reply opensearch-operator/api/v1/opensearch_types.go:50]: I planed to do it anyway to add configurations, will add it now

File: opensearch-operator/pkg/builders/cluster.go (hunk 4 above)
Comment (idanl21) [reply opensearch-operator/pkg/builders/cluster.go:199]: Probably it will not work, maybe we can add a guide, on how to download it and store it in a place where you can pull it. have another idea?
by the way, I think that problem will be also a problem with official OpenSearch plugins.

File: opensearch-operator/pkg/builders/cluster.go (hunk 1 above)
Comment (swoehrl-mw) [reply opensearch-operator/pkg/builders/cluster.go:900]: > the user should be already created? or we will create it as part of the flow?

I'd say the user is in charge of creating that extra monitoring user if they supply credentials. So assume it is already created.

File: opensearch-operator/examples/opensearch-cluster.yaml [hunk 5]
```diff
@@ -1,17 +1,17 @@
 apiVersion: opensearch.opster.io/v1
 kind: OpenSearchCluster
 metadata:
-  name: my-cluster
+  name: my-cluster1-3
```
Comment (swoehrl-mw) [inline opensearch-operator/examples/opensearch-cluster.yaml:4]: Why rename the example?

File: opensearch-operator/examples/opensearch-cluster.yaml [hunk 6]
```diff
@@ -1,17 +1,17 @@
...
   namespace: default
 spec:
   general:
-    version: 1.3.0
+    version: "1.3.0"
     httpPort: 9200
     vendor: opensearch
-    serviceName: my-cluster
-    pluginsList: ["repository-s3"," https://github.com/aiven/prometheus-exporter-plugin-for-opensearch/releases/download/1.3.0.0/prometheus-exporter-1.3.0.0.zip"]
+    serviceName: my-cluster1-3
```
Comment (swoehrl-mw) [inline opensearch-operator/examples/opensearch-cluster.yaml:11]: Same here, why rename?

File: docs/designs/crd.md [hunk 7]
```diff
@@ -401,3 +406,48 @@ Every NodePool is defining different Opensearch Nodes StatefulSet
...
+        <td>Define interval for scarping</td>
+        <td>false</td>
+        <td>30s</td>
+      </tr><tr>
+      </tr><tr>
+        <td><b>offlinePlugin</b></td>
+        <td>string</td>
+        <td>Define offline link to Aiven Plugin</td>
+        <td>false</td>
+        <td>https://github.com/aiven/prometheus-exporter-plugin-for-opensearch</td>
```
Comment (swoehrl-mw) [inline docs/designs/crd.md:458]: The example URL should be a complete path with versions and all as the operator expects it. Also missing is info that the plugin must match the opensearch version.

File: opensearch-operator/api/v1/opensearch_types.go [hunk 8]
```diff
@@ -85,11 +85,17 @@ type PVCSource struct {
 // ConfMgmt defines which additional services will be deployed
 type ConfMgmt struct {
 	AutoScaler  bool `json:"autoScaler,omitempty"`
-	Monitoring  bool `json:"monitoring,omitempty"`
 	VerUpdate   bool `json:"VerUpdate,omitempty"`
 	SmartScaler bool `json:"smartScaler,omitempty"`
 }
 
+type MonitoringStuck struct {
```
Comment (swoehrl-mw) [inline opensearch-operator/api/v1/opensearch_types.go:93]: Wierd name for the struct. Shouldn't this be something like `MonitoringConfig`?

File: opensearch-operator/api/v1/opensearch_types.go [hunk 9]
```diff
@@ -85,11 +85,17 @@ type PVCSource struct {
...
+	Enable         bool   `json:"enable,omitempty"`
+	MonitoringUser string `json:"monitoringUser,omitempty"`
+	Interval       string `json:"interval,omitempty"`
+	OfflinePlugin  string `json:"offlinePlugin,omitempty"`
```
Comment (swoehrl-mw) [inline opensearch-operator/api/v1/opensearch_types.go:97]: I would prefer this be named something like `PluginURL` to make it clearer whats inside. The offline part belongs in the description in the docs.

File: opensearch-operator/pkg/builders/cluster.go [hunk 10]
```diff
@@ -194,10 +194,15 @@ func NewSTSForNodePool(
 
 	var mainCommand []string
 	com := "./bin/opensearch-plugin install --batch"
-	if cr.Spec.General.Monitoring {
-		clusterVersion := cr.Spec.General.Version
-		url := "https://github.com/aiven/prometheus-exporter-plugin-for-opensearch/releases/download/" + clusterVersion + ".0/prometheus-exporter" + clusterVersion + ".0.zip"
-		cr.Spec.General.PluginsList = append(cr.Spec.General.PluginsList, url)
+	if cr.Spec.General.Monitoring.Enable {
+		if cr.Spec.General.Monitoring.OfflinePlugin == "" {
```
Comment (swoehrl-mw) [inline opensearch-operator/pkg/builders/cluster.go:199]: Condition is the wrong way. then block should be entered if `OfflinePlugin != ""`

File: opensearch-operator/api/v1/opensearch_types.go [hunk 11]
```diff
@@ -85,11 +85,17 @@ type PVCSource struct {
...
 type ConfMgmt struct {
 	AutoScaler  bool `json:"autoScaler,omitempty"`
-	Monitoring  bool `json:"monitoring,omitempty"`
 	VerUpdate   bool `json:"VerUpdate,omitempty"`
 	SmartScaler bool `json:"smartScaler,omitempty"`
 }
 
+type MonitoringStuck struct {
+	Enable         bool   `json:"enable,omitempty"`
+	MonitoringUser string `json:"monitoringUser,omitempty"`
```
Comment (swoehrl-mw) [inline opensearch-operator/api/v1/opensearch_types.go:95]: I'd prefer the user has the option to specify the name of a secret (via `corev1.LocalObjectReference`) that must contain username and password, and just default on the admin password secret if not specified.

File: docs/designs/crd.md [hunk 12]
```diff
@@ -408,3 +413,48 @@ Every NodePool is defining different Opensearch Nodes StatefulSet
...
+            <th>Type</th>
+            <th>Description</th>
+            <th>Required</th>
+            <th>default</th>
+        </tr>
+    </thead>
+    <tbody><tr>
+        <td><b>enable</b></td>
+        <td>bool</td>
+        <td>Define if to ernable monitoring for that cluster</td>
```
Comment (swoehrl-mw) [inline docs/designs/crd.md:437]:
```suggestion
        <td>Define if to enable monitoring for that cluster</td>
```

File: docs/userguide/main.md [hunk 13]
```diff
@@ -488,3 +488,26 @@ spec:
   roles:
   - sample-role
 ```
+
+
+## Adding Opensearch Monitoring to your cluster
+
+The operator allows you to install the Aiven monitoring for OpenSearch on your cluster as a build in feature (https://github.com/aiven/prometheus-exporter-plugin-for-opensearch)
```
Comment (swoehrl-mw) [inline docs/userguide/main.md:495]:
```suggestion
The operator allows you to install and enable the Aiven monitoring plugin for OpenSearch on your cluster as a built-in feature (https://github.com/aiven/prometheus-exporter-plugin-for-opensearch)
```

File: docs/userguide/main.md [hunk 14]
```diff
@@ -488,3 +488,26 @@ spec:
...
+That feature is needed Online connectivity to download the plugin. if you are working at offline ENV, please download and store where you have access that zip (https://github.com/aiven/prometheus-exporter-plugin-for-opensearch/releases/download/2.3.0.0/prometheus-exporter-2.3.0.0.zip).
```
Comment (swoehrl-mw) [inline docs/userguide/main.md:496]:
```suggestion
That feature needs internet connectivity to download the plugin. if you are working in a restricted environment, please 
download the plugin zip for your cluster version (example for 2.3.0: https://github.com/aiven/prometheus-exporter-plugin-for-opensearch/releases/download/2.3.0.0/prometheus-exporter-2.3.0.0.zip) and provide it at a location the operator can reach. Configure that URL as `pluginURL` in the monitoring config.
```

File: docs/userguide/main.md [hunk 15]
```diff
@@ -488,3 +488,26 @@ spec:
...
+after that please insert the new URL under offlinePlugin key.
```
Comment (swoehrl-mw) [inline docs/userguide/main.md:497]: Missing documentation on the monitoringUser. Should also mention the needed permissions if a custom opensearch user is used and that it defaults to the cluster admin user.

File: docs/userguide/main.md [hunk 16]
```diff
@@ -488,3 +488,26 @@ spec:
...
+metadata:
+  name: sample-urb
+spec:
+  opensearchCluster:
+    name: my-first-cluster
+    namespace: default
+  general:
+    monitoring:
+      enable: true
+      internval: 10s
```
Comment (swoehrl-mw) [inline docs/userguide/main.md:510]:
```suggestion
      interval: 30s
```

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]

File: opensearch-operator/controllers/suite_test.go [hunk 17]
```diff
@@ -85,13 +89,17 @@ var _ = BeforeSuite(func() {
...
+	err = prometheus.AddToScheme(scheme.Scheme)
+	Expect(err).NotTo(HaveOccurred())
 
 	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
 	Expect(err).NotTo(HaveOccurred())
 	Expect(k8sClient).NotTo(BeNil())
 
 	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
-		Scheme: scheme.Scheme,
+		Scheme:                 scheme.Scheme,
```
Comment (idanl21) [inline opensearch-operator/controllers/suite_test.go:100]: Hey @swoehrl-mw
can you please explain why you add the `MetricsBindAddress` and `HealthProbeBindAddress` ?

**Review by idanl21 — COMMENTED** [review]
Hey @swoehrl-mw, first of all thanks for the help, It's really not obvious.
i asked some questions , just for better understating in next time.

let's talk also about mergeing it and release new version :)

File: opensearch-operator/controllers/suite_test.go (hunk 17 above)
Comment (swoehrl-mw) [reply opensearch-operator/controllers/suite_test.go:100]: By adding both I can configure them with random ports so there are no collisions. I just copied that part from the other package so they behave the same and because I ran into a collision while testing locally.

File: opensearch-operator/api/v1/opensearch_types.go (hunk 11 above)
Comment (idanl21) [reply opensearch-operator/api/v1/opensearch_types.go:102]: needs to implement that logic, maybe we can declare it as a phase 2 feature and add it in the future so we will be able to release tomorrow .

File: docs/userguide/main.md (hunk 15 above)
Comment (idanl21) [reply docs/userguide/main.md:626]: about the perms, im not sure that i know which perms are needed for it..
will check that

File: opensearch-operator/pkg/reconcilers/cluster.go [hunk 18]
```diff
@@ -64,6 +63,18 @@ func (r *ClusterReconciler) Reconcile() (ctrl.Result, error) {
...
 	}
 
+	if r.instance.Spec.General.Monitoring.Enable {
+		serviceMonitor := builders.NewServiceMonitor(r.instance)
+		result.CombineErr(ctrl.SetControllerReference(r.instance, serviceMonitor, r.Client.Scheme()))
+		result.Combine(r.ReconcileResource(serviceMonitor, reconciler.StatePresent))
+
+	} else {
+		serviceMonitor := builders.NewServiceMonitor(r.instance)
+		resultMo, err := r.ReconcileResource(serviceMonitor, reconciler.StateCreated)
```
Comment (swoehrl-mw) [inline opensearch-operator/pkg/reconcilers/cluster.go:73]: You can just use `StateAbsent` and then don't need the extra check.

File: opensearch-operator/pkg/builders/cluster.go [hunk 19]
```diff
@@ -972,3 +984,67 @@ func DataNodesCount(ctx context.Context, k8sClient client.Client, cr *opsterv1.O
...
+					TargetPort:      nil,
+					Path:            "/_prometheus/metrics",
+					Interval:        scrape,
+					TLSConfig:       nil,
+					BearerTokenFile: "",
+					HonorLabels:     false,
+					BasicAuth: &prometheus.BasicAuth{
+						Username: corev1.SecretKeySelector{
+							LocalObjectReference: corev1.LocalObjectReference{Name: cr.Name + monitorUser + "-password"},
+							Key:                  monitorUser,
```
Comment (swoehrl-mw) [inline opensearch-operator/pkg/builders/cluster.go:1037]: This is a logic error. You cannot use the user as the key. Did you intend to use `username` as key?

File: opensearch-operator/pkg/builders/cluster.go [hunk 20]
```diff
@@ -972,3 +984,67 @@ func DataNodesCount(ctx context.Context, k8sClient client.Client, cr *opsterv1.O
...
+				{Port: "9200",
+					TargetPort:      nil,
+					Path:            "/_prometheus/metrics",
+					Interval:        scrape,
+					TLSConfig:       nil,
+					BearerTokenFile: "",
+					HonorLabels:     false,
+					BasicAuth: &prometheus.BasicAuth{
+						Username: corev1.SecretKeySelector{
+							LocalObjectReference: corev1.LocalObjectReference{Name: cr.Name + monitorUser + "-password"},
```
Comment (swoehrl-mw) [inline opensearch-operator/pkg/builders/cluster.go:1036]: How is the secret created?

File: opensearch-operator/controllers/cluster_test.go [hunk 21]
```diff
@@ -85,6 +87,12 @@ var _ = Describe("Cluster Reconciler", func() {
 			Expect(k8sClient.Create(context.Background(), &OpensearchCluster)).Should(Succeed())
 		})
 
+		It("should create a ServiceMonitor for the cluster", func() {
+			Eventually(func() error {
+				return k8sClient.Get(context.Background(), client.ObjectKey{Name: OpensearchCluster.Name + "-monitor", Namespace: OpensearchCluster.Namespace}, &prometheus.ServiceMonitor{})
+			}, timeout, interval).Should(Succeed())
```
Comment (swoehrl-mw) [inline opensearch-operator/controllers/cluster_test.go:93]: That test is too minimal. Please also check that the created ServiceMonitor is configured with the correct endpoints, scrape interval and check if the secret is created with the correct credentials.

File: opensearch-operator/pkg/builders/cluster.go (hunk 19 above)
Comment (idanl21) [reply opensearch-operator/pkg/builders/cluster.go:1037]: yeah, my bad, fixed

File: docs/designs/crd.md [hunk 22]
```diff
@@ -492,6 +501,34 @@ Every Keystore Value defines a secret to pull secrets from.
...
+        <td>Define interval for scraping</td>
+        <td>false</td>
+        <td>30s</td>
+      </tr><tr>
+      </tr><tr>
+        <td><b>PluginURL</b></td>
+        <td>string</td>
+        <td>Define offline link to Aiven Plugin</td>
+        <td>false</td>
+        <td>https://github.com/aiven/prometheus-exporter-plugin-for-opensearch</td>
```
Comment (swoehrl-mw) [inline docs/designs/crd.md:528]: The example URL should be a complete path with versions and all as the operator expects it. Also missing is info that the plugin must match the opensearch version.

File: docs/designs/crd.md [hunk 23]
```diff
@@ -492,6 +501,34 @@ Every Keystore Value defines a secret to pull secrets from.
             <th>default</th>
         </tr>
     </thead>
+    <tbody><tr>
```
Comment (swoehrl-mw) [inline docs/designs/crd.md:504]: This is off by a few lines. The table for the monitoring fields is currently under the keystore section, needs to be moved to the monitoring section.

File: docs/userguide/main.md [hunk 24]
```diff
@@ -617,3 +617,26 @@ spec:
...
+
+The operator allows you to install the Aiven monitoring for OpenSearch on your cluster as a build in feature (https://github.com/aiven/prometheus-exporter-plugin-for-opensearch)
+That feature is needed Online connectivity to download the plugin. if you are working at offline ENV, please download and store where you have access that zip (https://github.com/aiven/prometheus-exporter-plugin-for-opensearch/releases/download/2.3.0.0/prometheus-exporter-2.3.0.0.zip).
+after that please insert the new URL under offlinePlugin key.
+```yaml
+apiVersion: opensearch.opster.io/v1
+kind: OpensearchUserRoleBinding
+metadata:
+  name: sample-urb
+spec:
```
[the remark this answers is not in the document]
Comment (swoehrl-mw) [reply docs/userguide/main.md:632]: My comment is still valid, the yaml shown here is inconsistent and invalid.

File: opensearch-operator/api/v1/opensearch_types.go [hunk 25]
```diff
@@ -92,11 +93,17 @@ type PVCSource struct {
...
 	AutoScaler  bool `json:"autoScaler,omitempty"`
-	Monitoring  bool `json:"monitoring,omitempty"`
 	VerUpdate   bool `json:"VerUpdate,omitempty"`
 	SmartScaler bool `json:"smartScaler,omitempty"`
 }
 
+type MonitoringConfig struct {
+	Enable               bool   `json:"enable,omitempty"`
+	MonitoringUserSecret string `json:"monitoringUserSecret,omitempty"`
+	ScrapInterval        string `json:"scrapInterval,omitempty"`
```
Comment (swoehrl-mw) [inline opensearch-operator/api/v1/opensearch_types.go:103]: Typo.
```suggestion
	ScrapeInterval        string `json:"scrapeInterval,omitempty"`
```

File: opensearch-operator/config/crd/bases/crd-servicemonitor.yaml [hunk 26]
```diff
@@ -0,0 +1,632 @@
+apiVersion: apiextensions.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (swoehrl-mw) [reply opensearch-operator/config/crd/bases/crd-servicemonitor.yaml:1]: My earlier comment was not handled.

File: opensearch-operator/controllers/dashboard_test.go [hunk 27]
```diff
@@ -116,8 +112,7 @@ var _ = Describe("Dashboards Reconciler", func() {
 			}, timeout, interval).Should(BeTrue())
 		})
 		It("should set correct owner references", func() {
-			Expect(HasOwnerReference(&deploy, &OpensearchCluster)).To(BeTrue())
-			Expect(HasOwnerReference(&cm, &OpensearchCluster)).To(BeTrue())
```
Comment (swoehrl-mw) [inline opensearch-operator/controllers/dashboard_test.go:120]: Why was this line removed?

File: opensearch-operator/examples/opensearch-cluster.yaml [hunk 28]
```diff
@@ -5,13 +5,15 @@ metadata:
...
   general:
-    version: 1.3.0
+    version: "1.3.0"
     httpPort: 9200
     vendor: opensearch
     serviceName: my-cluster
-    pluginsList: ["repository-s3"," https://github.com/aiven/prometheus-exporter-plugin-for-opensearch/releases/download/1.3.0.0/prometheus-exporter-1.3.0.0.zip"]
+    monitoring:›
+     enable: true
+#    pluginsList: ["repository-s3"]
```
Comment (swoehrl-mw) [inline opensearch-operator/examples/opensearch-cluster.yaml:14]: This line belongs to general and not monitoring, needs to be moved up so not to confuse people.

File: opensearch-operator/examples/opensearch-cluster.yaml [hunk 29]
```diff
@@ -5,13 +5,15 @@ metadata:
   namespace: default
 spec:
   general:
-    version: 1.3.0
+    version: "1.3.0"
     httpPort: 9200
     vendor: opensearch
     serviceName: my-cluster
-    pluginsList: ["repository-s3"," https://github.com/aiven/prometheus-exporter-plugin-for-opensearch/releases/download/1.3.0.0/prometheus-exporter-1.3.0.0.zip"]
+    monitoring:›
```
Comment (swoehrl-mw) [inline opensearch-operator/examples/opensearch-cluster.yaml:12]: Typo.
```suggestion
    monitoring:
```

File: opensearch-operator/pkg/builders/dashboards.go [hunk 30]
```diff
@@ -169,15 +169,15 @@ func NewDashboardsConfigMapForCR(cr *opsterv1.OpenSearchCluster, name string, co
 	for _, key := range keys {
 		sb.WriteString(fmt.Sprintf("%s: %s\n", key, config[key]))
 	}
-	data := sb.String()
+	datas := sb.String()
```
Comment (swoehrl-mw) [inline opensearch-operator/pkg/builders/dashboards.go:172]: There is no plural of data.

File: opensearch-operator/pkg/builders/cluster.go [hunk 31]
```diff
@@ -972,3 +984,80 @@ func DataNodesCount(ctx context.Context, k8sClient client.Client, cr *opsterv1.O
...
+		cr.Spec.General.Monitoring.ScrapInterval = "30s"
+	}
+	user := v1.BasicAuth{}
+
+	monitorUser := cr.Spec.General.Monitoring.MonitoringUserSecret
+	if monitorUser == "" {
+		monitorUser = "admin"
+		user = v1.BasicAuth{
+			Username: corev1.SecretKeySelector{
+				LocalObjectReference: corev1.LocalObjectReference{Name: cr.Name + monitorUser + "-password"},
```
Comment (swoehrl-mw) [inline opensearch-operator/pkg/builders/cluster.go:1012]: Writing it like this is confusing as it is unclear where this secret is created.
Suggestion: Define variable `opensearchAdminSecret = cr.Name + "admin-secret` with a comment `// Use admin credentials if no separate monitoring user was defined` and use that.

File: opensearch-operator/pkg/builders/cluster.go [hunk 32]
```diff
@@ -972,3 +984,80 @@ func DataNodesCount(ctx context.Context, k8sClient client.Client, cr *opsterv1.O
...
+		Spec: v1.ServiceMonitorSpec{
+			JobLabel: cr.Name + "-monitor",
+			TargetLabels: []string{
+				cr.Name,
+			},
+			PodTargetLabels: []string{
+				cr.Name,
+			},
+			Endpoints: []v1.Endpoint{
+				{Port: "9200",
```
Comment (swoehrl-mw) [inline opensearch-operator/pkg/builders/cluster.go:1049]: Port cannot be hardcoded as it is configurable in the custom resource (´general.httpPort´).

File: opensearch-operator/controllers/cluster_test.go [hunk 33]
```diff
@@ -84,7 +84,22 @@ var _ = Describe("Cluster Reconciler", func() {
 		It("should apply the cluster instance successfully", func() {
 			Expect(k8sClient.Create(context.Background(), &OpensearchCluster)).Should(Succeed())
 		})
-
+		//	Sm := &prometheus.ServiceMonitor{}
```
Comment (swoehrl-mw) [inline opensearch-operator/controllers/cluster_test.go:87]: Why is this test commented out?
Also: There is no specific unittest for the monitoring feature.

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]

File: docs/designs/crd.md (hunk 22 above)
Comment (idanl21) [reply docs/designs/crd.md:528]: Cause that i dont want to give an example that related to 1 specific version, i gave the plugin repo.
The operator knows how to build the plugin with the needed version automatically

File: opensearch-operator/pkg/builders/cluster.go (hunk 32 above)
Comment (idanl21) [reply opensearch-operator/pkg/builders/cluster.go:1049]: fixed

File: docs/designs/crd.md (hunk 22 above)
Comment (swoehrl-mw) [reply docs/designs/crd.md:528]: But thats the point, if you supply a custom url the operator does not know how to build a complete URL with versions, it just takes the URL as-is. So giving just the base path will lead to errors. You should give an URL with placeholders for the opensearch version so the users know what to put in.

File: opensearch-operator/controllers/cluster_test.go (hunk 33 above)
Comment (idanl21) [reply opensearch-operator/controllers/cluster_test.go:87]: Sorry, changed and added more tests

File: docs/designs/crd.md (hunk 22 above)
Comment (idanl21) [reply docs/designs/crd.md:528]: ho ok, np, added <YOUR_CLUSTER_VERSION> to the example

**Comment by idanl21** [comment]
closing and opened another one cause of merge problems 
https://github.com/Opster/opensearch-k8s-operator/pull/415
